CONTRIBUTING.md

```diff
@@ -0,0 +1,47 @@
+# Contribution rules
+
+The following subsections address how to contribute to the Space ROS project.
+
+Both internal and external contributors should follow these rules when
+submitting contributions to Space ROS. These rules have been designed to
+simplify the process of assuring software and establishing compliance with
+requirements of space software.
+
+## Commits
+
+- All commits must reference the issue they address. *A consequence of this is
+  that no commits can be made unless there is a corresponding issue for them.*
+
+## Issues
+
+- All issues being addressed must be assigned a milestone (ideally, prior to
+  engaging in the work).
+
+## Milestones
+
+- All releases are tagged with their version number, which also corresponds to
+  a milestone.
+
+## Pull Requests
+
+- All changes are incorporated via pull requests (even changes by the core
+  team).
+
+- All PR merges introduce a separate merge commit (i.e., git merge --no-ff),
+  which Closes the issue that the PR addresses, as well as any other issues
+  that are fixed as a side effect, or that can no longer be reproduced after the
+  change.
+
+- There'll be a preference for PRs to address only one issue at a time, but
+  we'll be flexible especially for cases in which a fix addresses multiple
+  related issues at the same time (see point above).
+
+- All PRs must be approved by someone other than the committer.
+
+- All PRs must pass the tests.
```

Comment:
> - All PRs must pass the tests for the repository they are being committed to.

I'm ok with this. If you think it's confusing or ambiguous, let's disambiguate it.

> - PRs which add functionality should also add tests to test that functionality.

We should avoid _shoulds_. If this is not a must, then we're very likely to ignore it, especially in times of pressure when releases may be most crucial.

In the projects I've worked on, we have a separate testing plan, and that determines what tests are required to be implemented at all, how we keep records, etc.

I would still like to keep a rule like:

> - All PRs must be approved by someone other than the committer.

This is not just our team. For example, Google has AFAIK similar rules.

Reply:
Nevertheless, we can word that rule "All PRs must be approved by someone other than the committer." a bit differently, for example, to make it clear who those "others" can be.

Reply:
> "must include tests"

I agree with the idea of having this somewhere, but traditionally this has not been part of the development process.

There are other aspects of how we develop code that are normally covered by other guidelines or plans. Testing is one; another example is coding standards, which covers code documentation (you did not mention it but it'll help illustrate my point). It's really hard to capture the complexity of those matters in a short bulleted list like the one in this PR.

Instead of dictating here that new features should include tests, I think it would pay to think about what we need to test and when, and create a test plan that covers those matters. That would be a separate document, or at least a separate PR. Once we have that, we can say that "Code contributions must abide by our test plan".

Similarly, instead of trying to state here generic coding rules, like _every module or class must have documentation_, it's best to prepare a coding standard/guidelines, and later add to these contribution rules a generic statement like "Code contributions must conform with out coding standards".